framework/src/org/checkerframework/framework/type/treeannotator/TreeAnnotator.java

```diff
@@ -32,4 +33,20 @@ public TreeAnnotator(AnnotatedTypeFactory atypeFactory) {
     public Void visitMethod(MethodTree node, AnnotatedTypeMirror p) {
         return super.visitMethod(node, p);
     }
+
+    /**
+     * When overriding this method, getAnnotatedType on the left and right operands should only be
+     * called when absolutely necessary. Otherwise, the checker will be very slow on heavily nested
+     * binary trees. (For example, a + b + c + d + e + f + g + h.)
+     *
+     * <p>If a checker's performance is still too slow, the types of binary trees could be computed
+     * in a subclass of {@link org.checkerframework.framework.flow.CFTransfer}. When computing the
+     * types in a transfer, look up in the value in the store rather than the AnnotatedTypeFactory.
+     * Then this method should annotated binary trees with top so that the type applied in the
+     * transfer is always a subtype of the type the AnnotatedTypeFactory computes.
+     */
+    @Override
+    public Void visitBinary(BinaryTree node, AnnotatedTypeMirror mirror) {
```

Comment:
In the PropagationTreeAnnotator above you also override other methods and check whether the types are fully annotated. Should there be similar comments like this one for these other methods?